File: ops/static/schedules.js
// schedules.js — recurring claims, and when the agreement is up.
//
// Maintenance is one agreement spread over a year, not twelve claims
// someone typed. The renewal date is the point: an agreement that lapses
// unnoticed is revenue that simply stops, and a spreadsheet of twelve rows
// never tells you it is about to.
//
// Every recurring project arrived with its rows ALREADY TYPED, so a new
// schedule adopts them rather than generating over the top. Generate then
// fills only the gaps.

import { api, fmt, h, moneyInput, mount, stateMessage, toCents } from "./app.js";

// Survives the re-render that follows an action. Setting a notice and then
// rebuilding the screen wipes it -- which made `Adopt` look like it had
// done nothing at all, when what it had done was correctly report that
// there was nothing left to adopt.
let pending = null;

const STATE_LABEL = {
  overdue: "Overdue", due: "Due", future: "Upcoming", "no date set": "No date",
};

function field(label, control, hint) {
  const error = h("span", { class: "field-error" });
  const wrap = h("label", { class: "field" },
    h("span", { class: "field-label" }, label),
    control,
    hint ? h("span", { class: "field-hint" }, hint) : null,
    error);
  return { wrap, control, setError: (m) => {
    error.textContent = m || "";
    wrap.classList.toggle("has-error", !!m);
  } };
}

function scheduleForm({ schedule, projects, periods, onSaved }) {
  const editing = !!schedule;
  const s = schedule || {};

  const project = h("select", { "aria-label": "Project" },
    h("option", { value: "" }, "Select a project"),
    projects.map((p) => h("option", { value: String(p.id) },
      `${p.name} \u00b7 ${p.job_code}`)));
  const period = (selected) => h("select", { "aria-label": "Period" },
    periods.map((p) => h("option",
      { value: String(p.id), selected: p.id === selected },
      `${p.label}  ${p.fy_label}`)));

  const fields = {
    description: field("Description",
      h("input", { type: "text", value: s.description || "",
                   "aria-label": "Description" }),
      "Labels every claim this makes"),
    amount_cents: field("Amount each time",
      moneyInput(s.amount_cents ?? 0, { "aria-label": "Amount" }), "Ex-GST"),
    frequency: field("Frequency",
      h("select", { "aria-label": "Frequency" },
        ["monthly", "quarterly", "annual"].map((f) => h("option",
          { value: f, selected: f === s.frequency }, f)))),
    start_period_id: field("First month", period(s.start_period_id)),
    end_period_id: field("Last month", period(s.end_period_id)),
    renewal_date: field("Renewal date",
      h("input", { type: "date", value: s.renewal_date || "",
                   "aria-label": "Renewal date" }),
      "When the agreement itself is up"),
    renewal_notice_days: field("Notice",
      h("input", { type: "number", min: "0",
                   value: String(s.renewal_notice_days ?? 60),
                   "aria-label": "Notice days" }),
      "Days before it starts chasing"),
    renewal_note: field("Renewal note",
      h("textarea", { rows: 2, "aria-label": "Renewal note" },
        s.renewal_note || "")),
  };

  const summary = h("div", { class: "form-error", hidden: true });
  const save = h("button", { type: "button", class: "primary" },
    editing ? "Save" : "Create schedule");
  const cancel = h("button", { type: "button" }, "Cancel");

  const dialog = h("dialog", { class: "sheet", "aria-label": "Schedule" },
    h("div", { class: "sheet-head" },
      h("h2", null, editing ? s.description : "New schedule"),
      editing ? h("span", { class: "mono muted" }, s.project_name) : null),
    h("div", { class: "form-grid" },
      editing ? null : field("Project", project).wrap,
      Object.values(fields).map((f) => f.wrap)),
    summary,
    h("div", { class: "sheet-foot" },
      h("span", { class: "spacer" }), cancel, save));

  save.addEventListener("click", async () => {
    for (const f of Object.values(fields)) f.setError("");
    summary.hidden = true;
    const cents = toCents(fields.amount_cents.control.value);
    if (cents === null) {
      fields.amount_cents.setError("not an amount");
      return;
    }
    const payload = {
      description: fields.description.control.value.trim(),
      amount_cents: cents,
      frequency: fields.frequency.control.value,
      start_period_id: Number(fields.start_period_id.control.value),
      end_period_id: Number(fields.end_period_id.control.value),
      renewal_date: fields.renewal_date.control.value || null,
      renewal_notice_days: Number(fields.renewal_notice_days.control.value),
      renewal_note: fields.renewal_note.control.value.trim(),
    };
    if (!editing) {
      const chosen = Number(project.value);
      if (!chosen) {
        summary.textContent = "Pick a project";
        summary.hidden = false;
        return;
      }
      payload.project_id = chosen;
      const p = projects.find((x) => x.id === chosen);
      payload.customer_po_id = p && p.customer_po_id;
    }
    save.disabled = true;
    try {
      const result = editing
        ? await api("PATCH", `/api/schedules/${s.schedule_id}`, payload)
        : await api("POST", "/api/schedules", payload);
      dialog.close();
      // Say what adoption did. A schedule that silently attached twelve
      // existing claims looks identical to one that did nothing.
      const adopted = result.adopted && result.adopted.adopted;
      onSaved(adopted
        ? `Adopted ${adopted} existing claim${adopted === 1 ? "" : "s"}`
        : null);
    } catch (err) {
      const detail = err.detail;
      if (detail && typeof detail === "object") {
        let rest = [];
        for (const [key, message] of Object.entries(detail)) {
          if (fields[key]) fields[key].setError(message);
          else rest.push(`${key}: ${message}`);
        }
        if (rest.length) {
          summary.textContent = rest.join("; ");
          summary.hidden = false;
        }
      } else {
        summary.textContent = err.message;
        summary.hidden = false;
      }
    } finally {
      save.disabled = false;
    }
  });
  cancel.addEventListener("click", () => dialog.close());
  dialog.addEventListener("close", () => dialog.remove());
  document.body.appendChild(dialog);
  dialog.showModal();
  return dialog;
}

export async function render(root) {
  mount(root, stateMessage("Loading schedules", null, false));
  let data, periods, projects, me;
  try {
    [data, periods, projects, me] = await Promise.all([
      api("GET", "/api/schedules"),
      api("GET", "/api/periods"),
      api("GET", "/api/projects"),
      api("GET", "/api/me"),
    ]);
  } catch (err) {
    mount(root, stateMessage("Could not load schedules", err.message, true));
    return;
  }
  const canWrite = new Set(me.roles.map((r) => r.role)).has("operations");
  const notice = h("div", { class: "notice", hidden: true });
  if (pending) {
    notice.textContent = pending;
    notice.hidden = false;
    pending = null;
  }
  // `say` for messages that stay on this render; `report` for ones that
  // must outlive it.
  const say = (m) => { notice.textContent = m || ""; notice.hidden = !m; };
  const report = (m) => { pending = m; render(root); };
  const reload = (message) => report(message);

  async function act(schedule, what) {
    try {
      const result = await api(
        "POST", `/api/schedules/${schedule.schedule_id}/${what}`);
      if (what === "adopt") {
        const spare = (result.not_adopted || []).length;
        // A month holding two claims is worth saying out loud: the schedule
        // can only own one of them, so the other is either a genuine extra
        // or a duplicate somebody should look at.
        const leftover = spare
          ? ` \u2014 ${spare} claim${spare === 1 ? "" : "s"} left unattached, `
            + `in ${[...new Set(result.not_adopted.map((x) => x.period))].join(", ")}`
            + " (a month can hold more than one, this schedule owns one)"
          : "";
        report(result.adopted
          ? `${schedule.description}: adopted ${result.adopted} existing `
            + `claim${result.adopted === 1 ? "" : "s"}`
              + (result.differing.length
                 ? `, ${result.differing.length} at a different amount` : "")
              + leftover
          : `${schedule.description}: nothing to adopt \u2014 every period `
            + "is already attached to this schedule" + leftover);
      } else {
        report(result.created
          ? `${schedule.description}: created ${result.created} claim`
            + `${result.created === 1 ? "" : "s"} (${result.periods.join(", ")})`
          : `${schedule.description}: every period already covered`);
      }
    } catch (err) {
      say(err.message);
    }
  }

  // What each schedule actually covers, month by month. The endpoint has
  // existed since the module was written; without it the only way to learn
  // what Generate would do was to press it.
  const opened = new Set();

  async function detailRow(schedule, colspan) {
    const cell = h("td", { colspan: String(colspan) }, "Loading\u2026");
    const tr = h("tr", { class: "detail-row" }, cell);
    try {
      const preview = await api(
        "GET", `/api/schedules/${schedule.schedule_id}/preview`);
      mount(cell, h("div", { class: "periods" },
        preview.periods.map((p) => {
          const extra = p.others.length
            ? h("span", { class: "period-extra" },
                `+${p.others.length} other`)
            : null;
          return h("div", { class: `period is-${p.state}` },
            h("span", { class: "period-month" }, p.label),
            h("span", { class: "period-amount" },
              p.claim ? fmt.money(p.claim.amount_cents) : "\u2013"),
            h("span", { class: "period-state" },
              p.state === "mine" ? (p.claim.status)
                : p.state === "unattached" ? "not this schedule"
                : "no claim"),
            extra);
        })));
    } catch (err) {
      mount(cell, document.createTextNode(err.message));
    }
    return tr;
  }

  const rows = data.schedules;
  const chasing = rows.filter((r) => r.renewal_state === "overdue"
                                  || r.renewal_state === "due");

  // Toggling is DOM surgery rather than a re-render: rebuilding the screen
  // would close every other row and lose the scroll position.
  async function toggle(event, schedule) {
    const tr = event.currentTarget;
    const next = tr.nextElementSibling;
    if (next && next.classList.contains("detail-row")) {
      next.remove();
      opened.delete(schedule.schedule_id);
      tr.setAttribute("aria-expanded", "false");
      return;
    }
    opened.add(schedule.schedule_id);
    tr.setAttribute("aria-expanded", "true");
    tr.after(await detailRow(schedule, 10));
  }

  mount(root, h("div", { class: "content" },
    h("div", { class: "page-head" },
      h("h1", null, "Schedules"),
      h("span", { class: "eyebrow" }, "recurring claims"),
      h("span", { class: "spacer" }),
      canWrite
        ? h("button", { type: "button", class: "primary",
                        onclick: () => scheduleForm({
                          projects: projects.projects, periods: periods.periods,
                          onSaved: reload }) }, "New schedule")
        : null),
    notice,
    // Renewals first and loudest: it is the only thing here with a
    // deadline, and the reason the screen exists.
    chasing.length
      ? h("div", { class: "renewals" },
          h("h2", null, "Renewals"),
          chasing.map((r) => h("div",
            { class: `renewal is-${r.renewal_state}` },
            h("span", { class: "renewal-state" }, STATE_LABEL[r.renewal_state]),
            h("span", { class: "renewal-what" },
              `${r.project_name} \u00b7 ${r.description}`),
            h("span", { class: "mono" }, r.renewal_date || "\u2013"),
            h("span", { class: "muted" },
              r.days_until < 0
                ? `${Math.abs(r.days_until)} days ago`
                : `in ${r.days_until} days`),
            r.renewal_note ? h("span", { class: "muted" }, r.renewal_note) : null)))
      : null,
    rows.length
      ? h("div", { class: "table-wrap" },
          h("table", null,
            h("thead", null, h("tr", null,
              h("th", null, "Project"),
              h("th", null, "Description"),
              h("th", null, "Every"),
              h("th", null, "From"),
              h("th", null, "To"),
              h("th", { class: "num" }, "Each"),
              h("th", { class: "num" }, "Covered"),
              h("th", { class: "num" }, "Value"),
              h("th", null, "Renewal"),
              h("th", null, ""))),
            h("tbody", null, rows.map((r) => h("tr",
              { class: [r.is_active ? null : "muted", "clickable"]
                  .filter(Boolean).join(" "),
                tabindex: "0", role: "button",
                "aria-expanded": String(opened.has(r.schedule_id)),
                onclick: (e) => toggle(e, r),
                onkeydown: (e) => {
                  if (e.key === "Enter" || e.key === " ") {
                    e.preventDefault();
                    toggle(e, r);
                  }
                } },
              h("td", { class: "text-wide", title: r.project_name }, r.project_name),
              h("td", { class: "text", title: r.description }, r.description),
              h("td", { class: "mono" }, r.frequency),
              h("td", { class: "mono" }, r.start_label),
              h("td", { class: "mono" }, r.end_label),
              h("td", { class: "num" }, fmt.money(r.amount_cents)),
              // A fraction, not a count: complete or not is the question.
              h("td", { class: r.generated_count >= r.expected_count
                               ? "num" : "num flagged-cell" },
                `${r.generated_count} / ${r.expected_count}`),
              h("td", { class: "num" }, fmt.moneyDash(r.generated_cents)),
              h("td", { class: "mono" }, r.renewal_date || "\u2013"),
              h("td", { onclick: (e) => e.stopPropagation() },
                canWrite && r.is_active
                ? h("span", { class: "row-actions" },
                    h("button", { type: "button",
                                  title: "Attach claims that already exist",
                                  onclick: () => act(r, "adopt") }, "Adopt"),
                    h("button", { type: "button",
                                  title: "Create the claims that do not exist yet",
                                  onclick: () => act(r, "generate") }, "Generate"),
                    h("button", { type: "button",
                                  onclick: () => scheduleForm({
                                    schedule: r, projects: projects.projects,
                                    periods: periods.periods, onSaved: reload }) },
                      "Edit"))
                : null)))))) 
      : stateMessage("No schedules yet",
          "Maintenance and SLA work belongs here: one agreement, not twelve "
          + "rows a year.", false)));
}

File: ops/modules/schedules.py
"""Schedules module — recurring claims and renewals (STP-2).

Maintenance is one agreement spread over a year, not twelve claims someone
typed. `36 Wellington` is $22,689 as twelve payments of $1,890.75; entering
those by hand is the monthly copy-forward ritual in another guise, and next
year is twelve more.

Two operations, and the order matters:

  ADOPT     attach claims that already exist to the schedule describing them
  GENERATE  create the ones that do not exist yet

Every recurring project in the register arrived with its twelve rows already
typed, so generating first would double the year. Adopt is therefore offered
first and generation reports how many periods were already covered.

The renewal date is the point of the whole thing. A maintenance agreement
that lapses unnoticed is revenue that simply stops, and nothing in a
spreadsheet of twelve rows tells you it is about to.
"""

from typing import Any

from ops.db import Db
from ops.http_util import HttpError, Router

ROLE_READ = "viewer"
ROLE_WRITE = "operations"

FREQUENCIES = ("monthly", "quarterly", "annual")


def entity_ids(user: dict[str, Any]) -> list[int]:
    return sorted({r["entity_id"] for r in user["roles"]})


def validate(db: Db, payload: dict[str, Any], entity_id: int,
             existing: dict[str, Any] | None = None):
    errors: dict[str, str] = {}
    fields: dict[str, Any] = {}
    creating = existing is None

    if creating:
        project = db.query_one(
            "SELECT id, entity_id FROM project WHERE id = ?",
            (payload.get("project_id"),))
        if project is None or project["entity_id"] != entity_id:
            errors["project_id"] = "required; a project on this entity"
        else:
            fields["project_id"] = project["id"]
            po = db.query_one(
                "SELECT id FROM customer_po WHERE id = ? AND project_id = ?",
                (payload.get("customer_po_id"), project["id"]))
            if po is None:
                errors["customer_po_id"] = "required; which PO this bills against"
            else:
                fields["customer_po_id"] = po["id"]

    if creating or "description" in payload:
        text = (payload.get("description") or "").strip()
        if not text:
            # It becomes the detail on every claim it makes, so a blank one
            # produces a year of unlabelled rows.
            errors["description"] = "required; it labels every claim this makes"
        else:
            fields["description"] = text

    if creating or "amount_cents" in payload:
        try:
            amount = int(payload.get("amount_cents", 0))
        except (TypeError, ValueError):
            amount = 0
        if amount <= 0:
            errors["amount_cents"] = "required; the amount per occurrence"
        else:
            fields["amount_cents"] = amount

    if creating or "frequency" in payload:
        freq = (payload.get("frequency") or "").strip()
        if freq not in FREQUENCIES:
            errors["frequency"] = f"must be one of {', '.join(FREQUENCIES)}"
        else:
            fields["frequency"] = freq

    for key in ("start_period_id", "end_period_id"):
        if creating or key in payload:
            row = db.query_one("SELECT id, month_start FROM period WHERE id = ?",
                               (payload.get(key),))
            if row is None:
                errors[key] = "required"
            else:
                fields[key] = row["id"]
    if "start_period_id" in fields and "end_period_id" in fields:
        span = db.query_one(
            """SELECT (SELECT month_start FROM period WHERE id = ?) AS start,
                      (SELECT month_start FROM period WHERE id = ?) AS finish""",
            (fields["start_period_id"], fields["end_period_id"]))
        if span and span["start"] > span["finish"]:
            errors["end_period_id"] = "the schedule would end before it starts"

    if "renewal_date" in payload:
        fields["renewal_date"] = (payload.get("renewal_date") or "").strip() or None
    if "renewal_notice_days" in payload:
        try:
            fields["renewal_notice_days"] = max(0, int(payload["renewal_notice_days"]))
        except (TypeError, ValueError):
            errors["renewal_notice_days"] = "must be a whole number of days"
    if "renewal_note" in payload:
        fields["renewal_note"] = (payload.get("renewal_note") or "").strip() or None
    if "is_active" in payload:
        fields["is_active"] = 1 if payload["is_active"] else 0

    return fields, errors


def register(router: Router, db: Db) -> None:

    def scoped(user):
        ids = entity_ids(user)
        if not ids:
            raise HttpError(403, "no entity access")
        return ids

    def owned(user, schedule_id):
        ids = scoped(user)
        row = db.query_one(
            "SELECT * FROM claim_schedule WHERE id = ?", (schedule_id,))
        if row is None or row["entity_id"] not in ids:
            raise HttpError(404, "not found")
        return row

    @router.route("/api/schedules", role=ROLE_READ)
    def list_schedules(handler, user):
        ids = scoped(user)
        marks = ",".join("?" * len(ids))
        rows = db.query(
            f"""SELECT c.*,
                       (SELECT COUNT(*) FROM claim_line cl
                        WHERE cl.schedule_id = c.schedule_id) AS claim_count
                FROM v_upcoming_renewals c
                WHERE c.entity_id IN ({marks})
                ORDER BY c.days_until IS NULL, c.days_until""", tuple(ids))
        # Inactive schedules are excluded from the renewals view; list them
        # too, or a schedule someone switched off becomes unreachable.
        inactive = db.query(
            f"""SELECT * FROM v_schedule_coverage
                WHERE entity_id IN ({marks}) AND is_active = 0""", tuple(ids))
        out = rows + inactive
        # How many periods the schedule COVERS, so the count can be read as
        # a fraction. "12 claims" says nothing about whether that is all of
        # them; "12 / 12" says the year is complete and "6 / 12" says it is
        # not, which is the only reason to press Generate.
        for row in out:
            row["expected_count"] = len(db.schedule_periods(row["schedule_id"]))
        return 200, {"schedules": out, "frequencies": list(FREQUENCIES)}

    @router.route("/api/schedules", role=ROLE_WRITE, method="POST")
    def create_schedule(handler, user):
        payload = handler.read_json()
        ids = scoped(user)
        entity_id = payload.get("entity_id") or ids[0]
        if entity_id not in ids:
            raise HttpError(403, "no access to that entity")
        fields, errors = validate(db, payload, entity_id)
        if errors:
            raise HttpError(400, "validation failed", errors)
        fields["entity_id"] = entity_id
        schedule = db.create_schedule(fields, user["id"])
        # Adopt immediately: the claims almost certainly already exist, and
        # a schedule that appears to cover nothing invites someone to press
        # Generate and double the year.
        adopted = db.adopt_claims_into_schedule(schedule["id"], user["id"])
        return 201, {"schedule": schedule, "adopted": adopted}

    @router.route("/api/schedules/{schedule_id}", role=ROLE_WRITE,
                  method="PATCH")
    def update_schedule(handler, user, schedule_id):
        row = owned(user, schedule_id)
        payload = handler.read_json()
        unknown = [k for k in payload if k not in db.SCHEDULE_MUTABLE]
        if unknown:
            raise HttpError(400, "validation failed",
                            {k: "not an editable field" for k in unknown})
        fields, errors = validate(db, payload, row["entity_id"],
                                  existing=dict(row))
        if errors:
            raise HttpError(400, "validation failed", errors)
        result = db.update_schedule(row["id"], fields, user["id"])
        if result is None:
            raise HttpError(404, "not found")
        return 200, result

    @router.route("/api/schedules/{schedule_id}/adopt", role=ROLE_WRITE,
                  method="POST")
    def adopt(handler, user, schedule_id):
        row = owned(user, schedule_id)
        return 200, db.adopt_claims_into_schedule(row["id"], user["id"])

    @router.route("/api/schedules/{schedule_id}/generate", role=ROLE_WRITE,
                  method="POST")
    def generate(handler, user, schedule_id):
        row = owned(user, schedule_id)
        if not row["is_active"]:
            raise HttpError(409, "this schedule is not active")
        return 200, db.generate_schedule_claims(row["id"], user["id"])

    @router.route("/api/schedules/{schedule_id}/preview", role=ROLE_READ)
    def preview(handler, user, schedule_id):
        """Which months it covers, and what each already holds. Pressing
        Generate should never be the way to find out what it will do."""
        row = owned(user, schedule_id)
        out = []
        for period in db.schedule_periods(row["id"]):
            claims = db.query(
                """SELECT id, amount_cents, schedule_id, status, detail,
                          invoice_number
                   FROM claim_line
                   WHERE project_id = ? AND period_id = ?
                     AND is_opening_balance = 0 ORDER BY id""",
                (row["project_id"], period["id"]))
            mine = next((c for c in claims if c["schedule_id"] == row["id"]), None)
            # Every claim in the month, not just the one the schedule owns.
            # A period showing one claim when it holds two is how the Jul-26
            # surprise happened in the first place.
            others = [c for c in claims if c["schedule_id"] != row["id"]]
            out.append({
                "period_id": period["id"], "label": period["label"],
                "claim": dict(mine) if mine else None,
                "others": [dict(c) for c in others],
                "state": ("mine" if mine
                          else "unattached" if others else "missing"),
            })
        return 200, {"periods": out, "amount_cents": row["amount_cents"],
                     "description": row["description"]}
